### DISCUSSION TITLE
Doing DDD with Nuxt - automatized

### DISCUSSION THREAD

**Discussion by r9mp** [nuxt/nuxt#10071] [Questions (Nuxt 2)]
Hi there!

I saw at the Nuxt Nation a conf speaking about DDD with Nuxt.
Here is the repository: https://github.com/areindl/nuxt-structure-with-modules

Areindl is using modules, which is great. But the way he does it does not convince me as we are loosing a part of what makes Nuxt great: auto import of ressources (plugins, components, layouts, routes and store).

Do you know if there is some kind of module or hook (like for components: `components:dirs`) which exist to achieve that?

**Comment by alexcroox** [comment]
@r9mp did you figure this out? I'm looking to implement a custom module too but am surprised I'm having to look at third party auto routing libraries such as https://github.com/ktsn/vue-auto-routing when I'm inside a Nuxt project...

@areindl @pi0 are we right in saying if you create a custom module you can't utilise any of the Nuxt auto routing etc magic?

**Comment by r9mp** [comment]
@alexcroox I did not figured it out with an easy solution.
I made few tests, and it seems to be possible with third-party libraries (as you said).
But I stopped as it was time consuming and much easier to just call extendRoutes or directly create the store though Vue.
For me this should be avalaible through Nuxt hooks in the same way that the "component:dirs" hook (which is awesome and really simple to use).

**Comment by alexcroox** [comment]
@r9mp tackled this morning. Found in the Nuxt source code a `createRoute` function in @nuxt/utils and was able to set this up for dynamic routes for a custom module:

```
import { createRoutes } from '@nuxt/utils'
import glob from 'glob'
import path from 'path'

const moduleName = 'inventory'

// moduleOptions can pass config params from nuxt.config.js
module.exports = async function registerModule(moduleOptions) {
  // Automatically register all components of the module
  this.nuxt.hook('components:dirs', directories => {
    // Add ./components dir to the list
    directories.push({
      path: path.resolve(__dirname, 'components'),
      pathPrefix: false
    })
  })

  // Add module middleware
  this.nuxt.hook('build:templates', ({ templateVars }) => {
    templateVars.middleware.push({
      name: 'inventory-default',
      src: path.resolve(__dirname, 'middleware/inventory-default.js'),
      fileName: path.join('inventory-default.js')
    })

    templateVars.router.middleware.push('inventory-default')
  })

  const moduleRoutes = await generateModuleRoutes()

  // Add module routes to global router list
  this.extendRoutes(routes => {
    moduleRoutes.forEach(route => {
      routes.push(route)
    })
  })
}

function generateModuleRoutes() {
  return new Promise((resolve, reject) => {
    // Auto generate routes from custom pages directory
    glob(path.resolve(__dirname, 'pages/**/*.vue'), (error, files) => {
      if (error) {
        console.error(error)
        reject(error)
      }

      let moduleRoutes = createRoutes({
        files,
        srcDir: path.resolve(__dirname, 'pages'),
        pagesDir: path.resolve(__dirname, 'pages')
      })

      // Add our module name to the front of name and paths
      moduleRoutes = moduleRoutes.map(route => {
        route.name = `${moduleName}-${route.name}`
        route.path = `/${moduleName}${route.path}`
        return route
      })

      resolve(moduleRoutes)
    })
  })
}
```

**Reply by r9mp** [reply]
@alexcroox I did not had the routes part, and this seems really really cool! I'll definitively try it soon. Thanks a lot!
I had troubles with the store as well, but I'll check in the NuxtJS source code as you did when I'll retry.
Thanks again :)
